Source/Urho3D/Network/Network.cpp

```diff
@@ -200,40 +326,69 @@ void Network::ClientDisconnected(kNet::MessageConnection* connection)
     }
 }
 
+void Network::SetDiscoveryBeacon(const VariantMap& data)
+{
+    VectorBuffer buffer;
+    buffer.WriteVariantMap(data);
+    if (buffer.GetSize() > 400)
+        URHO3D_LOGERROR("Discovery beacon of size: " + String(buffer.GetSize()) + " bytes is too large, modify MAX_OFFLINE_DATA_LENGTH in RakNet or reduce size");
+    rakPeer_->SetOfflinePingResponse((const char*)buffer.GetData(), buffer.GetSize());
+}
+
+void Network::DiscoverHosts(unsigned port)
+{
+    // JSandusky: Contrary to the manual, we actually do have to perform Startup first before we can Ping
+    if (rakPeer_ && !rakPeer_->IsActive())
+    {
+        SLNet::SocketDescriptor socket;
+        rakPeer_->Startup(1, &socket, 1);
+    }
+    rakPeer_->Ping("255.255.255.255", port, false);
+}
+
+void Network::SetPassword(const String& password)
+{
+    rakPeer_->SetIncomingPassword(password.CString(), password.Length());
+    password_ = password;
+}
+
 bool Network::Connect(const String& address, unsigned short port, Scene* scene, const VariantMap& identity)
 {
     URHO3D_PROFILE(Connect);
 
-    // If a previous connection already exists, disconnect it and wait for some time for the connection to terminate

```

Comment:
Who sends `OnServerDisconnected` in AFTER version if `Connect` is called while connection is active?